### PR TITLE
treat known akamai responses as expected

### PR DESCRIPTION
**Pull request by ohnorobo**
Gave a different outcome to responses that are both from a known Akamai server header and send `x_on_this_server`. This is an attempt to start with the most conservative change we can make here that will still cause a large improvement in precision.

I went with giving the Akamai responses their own named outcome `expected/trusted_host` and treating them as expected (not just null) in the expected counts. This means they show up in the totals for the pivot table.

[test dashboard](https://datastudio.google.com/c/u/0/reporting/40dc493c-9974-415c-93ec-bab04de93b6d/page/T33gB) (I'm also experimenting in this dashboard with changing the outcome colors to match our sense of how unexpected something is. green=good, red=bad, yellow=maybe. Let me know how you like it.)

Example view of the UAE (surfaces blocking on AS5384 instead of being overwhelmed by Akamai errors)
Bcy5scVzsPhpooA

Example view of the US: (only shows the low level of non-`x-on-this-server` errors from Akamai
5697ZkE3JetihBw

### CODE REVIEW

File: table/queries/merged_reduced_scans.sql
Comment (fortuna) [inline table/queries/merged_reduced_scans.sql:176]: I believe it's time to update how we output expected results.
It would be helpful to make it clearer what's expected.
In this case, we could have `expected/matches_control` and `expected/trusted_host:akamai`
The unexpected_count formula could use `STARTS_WITH('expected/')`

File: table/queries/merged_reduced_scans.sql
Comment (fortuna) [inline table/queries/merged_reduced_scans.sql:34]: Won't this `content/mismatch` take precedence over the Akamai test?

File: table/queries/merged_reduced_scans.sql
Comment (fortuna) [inline table/queries/merged_reduced_scans.sql:116]: I believe there's no need to concatenate here. That may make the query significantly slower and harder to read.

You should be able to do something like 
```sql
select REGEXP_EXTRACT(header, "Server: (.*?)")
from UNNEST(received_headers) as header
where STARTS_WITH(header, "Server: ") LIMIT 1
```

Example: http://google3/jigsaw/firehook/metrics/update_intra_windowed_stats.sql;l=4;rcl=323798496

See https://cloud.google.com/bigquery/docs/reference/standard-sql/arrays#scanning_for_values_that_satisfy_a_condition

File: table/queries/merged_reduced_scans.sql
Comment (fortuna) [inline table/queries/merged_reduced_scans.sql:37]: What makes us trust the response is the fact that it came from Akamai. It wasn't injected.
That's true for all domains, hosted or not hosted on Akamai.

I believe it would be simpler to understand and explain if we just mark all tests that have the Akamai headers, and ignore the body pattern.

More importantly, we should prevent "expected" Akamai results from biasing the numbers when an AS happens to have Akamai, but it's not in the org id. Currently you will just remove the previously "unexpected" results.

We should also be more transparent. Let's output what the server is. The outcome could be `expected/trusted_host:akamai`

Comment (fortuna) [reply table/queries/merged_reduced_scans.sql:37]: One other thing that you don't need to address now: it's misleading to call the pattern id blockpage_id, when it isn't a block page. If the ids will include things that are not blockpages, we should rename it.

**Review by fortuna — CHANGES_REQUESTED** [review]

File: table/queries/merged_reduced_scans.sql
Comment (fortuna) [reply table/queries/merged_reduced_scans.sql:37]: While at it, we might as well add Cloudflare. I think they use `Server: cloudflare`.
Another reason to always return the known host outcome is that it won't change depending on whether DDoS protection kicks in.

**Comment by fortuna** [comment]
By the way, I do like the color idea. It will help visualize what's expected or not.

File: table/queries/merged_reduced_scans.sql
Comment (fortuna) [reply table/queries/merged_reduced_scans.sql:37]: Example of a Cloudflare page and its headers: https://explorer.ooni.org/measurement/20210817T073835Z_webconnectivity_KH_131178_n1_C08y4tgoO0sgboxv?input=http%3A%2F%2Flgbt.foundation%2F

Comment (ohnorobo) [reply table/queries/merged_reduced_scans.sql:37]: I'm specifically not trying to trust every server that returns an `AkamaiGHost` or `GHost` header. Those servers return a [bunch of other](https://datastudio.google.com/c/u/0/reporting/55e6ab82-8138-4ccc-a41f-66e6fb33d126/page/p_xd8o7a52mc?params=%7B%22df26%22:%22include%25EE%2580%25800%25EE%2580%2580IN%25EE%2580%2580HTTP%25EE%2580%2580HTTPS%22,%22df24%22:%22include%25EE%2580%25800%25EE%2580%2580IN%25EE%2580%2580AkamaiGHost%25EE%2580%2580GHost%22,%22df12%22:%22ORexclude%25EE%2580%25800%25EE%2580%2580IN%25EE%2580%2580content%252Fknown_response%253Ax_on_this_server%25EE%2580%2581include%25EE%2580%25803%25EE%2580%2580NU%25EE%2580%2582%22%7D) error patterns beyond just `x_on_this_server` including many unknown 3/400 errors. Many of those could be false positives but I haven't analysed them in any detail, and I wouldn't want to filter them out of the dashboard without looking more closely.

The point of this change is to do the most conservative thing (filtering out only the union of akamai server header and `x_on_this_server`) that we nevertheless know will make a large dashboard improvement.

Comment (fortuna) [reply table/queries/merged_reduced_scans.sql:37]: What I'm trying to get at is a simplified methodology that is easier to explain, understand and maintain.
It's a lot easier to say we are marking all Akamai responses as trusted_host:Akamai than saying we mark as that for some list of patterns that were arbitrarily selected and is not comprehensive. It's also a lot of work to vet all the patterns.

The assumption of my proposal is that censorship equipment will not try to impersonate other servers. That's not known to happen, and a mistake would treat sites as not blocked, so it's relatively safe to assume.

By marking all Akamai responses as expected we:

1. simplify the methodology. A simple methodology is important for people using the dashboard or trying to reproduce it. I appreciate the precaution to change less results, but I'd argue that a change that makes the methodology more complex is less conservative. It's more things to consider.
1. can filter out all of Akamai to remove bias, which is not possible if you are only marking some of the akamai results.
1. increase the coverage of errors that we mark as expected and the coverage is actually explainable as opposed to "some patterns we picked".
1. covers extra cases like the server successfully returning the actual site content, but not matching the control for some reason (e.g. regional content). You won't have patterns for those, other than the Server header or something else that says it's Akamai.

Comment (fortuna) [reply table/queries/merged_reduced_scans.sql:37]: As a side note, I'd prefer to not introduce any pattern that is based on the site content to the methodology, since that's really hard to scale and maintain.

Patterns for block pages are OK, since they are well known and obvious, and it's `O(censors)`, not `O(sites)`.
Patterns for hosting providers are ok too, since `num_hosting_providers` is much less than `num_sites`.

Comment (ramakrishnansr) [reply table/queries/merged_reduced_scans.sql:37]: First off, I agree with making the outcome be `expected/trusted_response:akamai` or `expected/trusted_response:pattern_id`, which makes it easier for users to connect with what we're matching to. Ideally, we should have all the patterns that we match against (be it response header or body) somewhere in this repository (or online) so that people can see what we're matching to. 

On whether we should mark all akamai responses as expected, I would say yes. Like I discussed, I think it's better to be conservative, and give users the default view of what we see (even if it is a censor, it's still an akamai page), and then if we get some knowledge of a pattern being incorrectly considered as expected, we can change it. But that should be the manual part, we should automatically mark these cases as expected. 

Also, I believe it is a really reasonable approach to match with the body patterns as well. There are plenty of cases where the server header is empty or contains different values, but the body still contains a very clear sign of a response from a hosting provider. We can find this by looking at how frequently a particular pattern appears. We shouldn't reduce the accuracy of the dashboard just because it is complicated to explain. I agree it may be a little hard to scale, but we've managed to have good coverage over 3 years, and the churn isn't as much because we consistently use the same IP addresses for measurements over time. 

Best approach to me seems: Have a public file of hosting provider responses which we consider as trusted, and match against all of them to conservatively mark as many expected responses as we can. This should be easy know for users to reproduce.

Comment (fortuna) [reply table/queries/merged_reduced_scans.sql:37]: I can see some patterns being useful, like the Cloudflare DDoS protection page. However, I'm curious to see cases where we can't classify looking at the headers alone. It's also safer to have patterns that are signature of specific hosting providers than generic patterns like `x_on_this_server`. However, this is a partial classification and it would mean biased numbers if we remove the CDN results that we know.
It's a tradeoff, and it will be interesting to explore more when we find such examples.

File: table/queries/merged_reduced_scans.sql
Comment (ohnorobo) [reply table/queries/merged_reduced_scans.sql:37]: It wasn't in my data, but it might in the time period where we're missing hyperquack v2 error output. Moved it above.

File: table/queries/merged_reduced_scans.sql
Comment (ohnorobo) [reply table/queries/merged_reduced_scans.sql]: Changed to 

`expected/match`
and
`expected/trusted_host:akamai`

File: table/queries/merged_reduced_scans.sql
Comment (ohnorobo) [reply table/queries/merged_reduced_scans.sql]: done

File: table/queries/merged_reduced_scans.sql
Comment (ohnorobo) [reply table/queries/merged_reduced_scans.sql]: Switched this to only match on the server pattern, not `x_on_this_server`.

I'm holding off on adding other hosts like cloudflare for now.

**Review by fortuna — APPROVED** [review]
Looks great. Thanks Sarah!

File: table/queries/merged_reduced_scans.sql
Comment (ramakrishnansr) [reply table/queries/merged_reduced_scans.sql]: I'm wondering why we're only marking Akamai and not the other obvious cases of expected responses? I agree that the majority of expected responses failing the template are Akamai, but there is a non-significant number of others as well that can be quite easily removed with the same approach.

Comment (ramakrishnansr) [reply table/queries/merged_reduced_scans.sql]: I'm worried that we bias the results towards other cloud providers now.

Comment (ohnorobo) [reply table/queries/merged_reduced_scans.sql]: Sorry, didn't see this comment before I merged. 

I wanted to just focus on akamai in this PR since we know the akamai precision increase is so good. Obviously we do want to filter out other expected responses, (I can already see cloudflare in the data) but deciding exactly what to filter/how to split that between the existing patterns and adding stuff to this query, requires some more thought.

**Pull request merged**
